### DISCUSSION TITLE
MVP, Geospatial and the Path Forward FY22

### DISCUSSION THREAD

**Discussion by Brennan1994** [HydrologicEngineeringCenter/HEC-FDA#83] [Q&A] [answered]
We have a meeting Sept. 20th about the path forward for FY22, and the MVP and Geospatial is going to be important, I wanted to start this thread to start brainstorming what that plan needs to be. Compiling some important points from our other discussions available here, and adding some notes about other discussions that happened outside github. 

Minimally Viable Product: 
- Differentiated the Guidance Conforming Product from the Minimally Viable Product. Guidance Conforming Product definition available here: https://www.hec.usace.army.mil/confluence/display/FDA/FDA+2.0+Requirements
- Will's proposal of a minimally viable product 
> Unregulated Flow Frequency -> Rating Curve with uncertainty ->User defined Aggregated Stage Damage Curve -> EAD is a minimally viable product, and could be tested via the model logic with no V or VM interaction.
Next we would want to connect it to a VM and test.
Then build a simple UI to trigger the compute from the VM down to the Model.

Geospatial Path Forward:
I laid out three alternatives summarized here, we agreed tentatively on Alt 1, Alt 4 and 5 are things we've discussed offline since

1. We can continue as is: referencing the specific set of .dll files that we need to run FDA as previously designed. This is already in place and working. It is not upgradable: we don't have source code for these libraries. No capability to fix bugs. Right now these .dlls are just stored in our repo under the lib folder. I could theoretically put these into NuGet Packages and host them on Nexus, which would make this infrastructure similar to alternative 2, except without source code.

2. We can upgrade to the latest version of LifeSims Geospatial libraries and package them as NuGet packages. With this method we're basically getting one big upgrade to whatever the latest feature additions/bug fixes are, then we're in a similar spot as before, except that I would reference as NuGet packages. With this path, we could try to keep up with LifeSim's development and I could regularly upgrade our dependencies by pulling latest and generating a new Nugets. It will likely require somewhat regular assistance from the LifeSim developer to keep everything building, and keep up with API changes.

3. We could fork off all LifeSim's code and maintain it ourselves. We'd then have full control for upgrades and bug fixes, but no assistance from the developer, as they'd no longer be his libraries. We'd be in charge of a lot of code with this alternative we have little no experience with.

4. Offline, we discussed RASMapper as a viable alternative to LifeSim GIS as our mapping tool. RASMapper would be beneficial in that we have the RAS Team in house, and we know it's supported. There's some question as to whether the RAS team has time to help us implement their tools, Talking with the RAS Team, I believe part of the appeal of helping us would be building inhouse skills, so it might not be in socially wise of us  to ask a contractor to take on this task, who will be joining the team at some point. 

5. Offline, we discussed not including a Map window at all. This gets back to the MVP. If we accept the proposed MVP above, it seems a map window is certainly not needed. If we reference the Guidance Conforming Product, the only place a map window shows up in our requirements is in the user QA/QC Requirements section, where we identify the need to view terrain data and depth/WSE rasters. In my interpretation of our requirements, we're not on the hook for a map window at all, though from a user perspective, the benefit is significant. 

I believe we're on the hook for providing a well thought out recommendation to our friends at our meeting, and would like to start fleshing that out among us here so we're ready to answer questions and express a consistent opinion when the time comes.

**Comment by Brennan1994** [comment]
I propose we lean on alternative 5, and point our initial, guidance conforming, to the districts release to be a map window-less product. Immediately someone will ask for a map, and we can provide that in 2.1 or 3.0. I think that meats all our needs in the most attainable way. We can process hydraulic data by ingesting hydraulics in the format of the dean method, and QAQC can be done in the same way it's been done in FDA 1.x.  

I think that also puts the least strain on our team, and we can focus our energies and resources on the more critical features.  This gets us off the hook for a lot of work integrating a map library. If we cruise through the rest of this development and find ourselves bored, I suggest we take the path of using RASMapper and pivot into alternative 4, but I don't foresee that happening, so I don't think that's really a decision we need to make now.

**Comment by HenryGeorgist** [comment]
To be clear about my intent for an MVP, I see it as a simple way for us to verify things “work”.  I anticipate we will have many mvps along the way. For instance, @rnugent3 added uncertainty to the mix - that is a good next logical, useful step. My hope would be that the mvp cycle would be short, for instance i will be sad if this first mvp takes more than two weeks. We need to be careful about pursuing perfection when our solution is good enough for most.

I agree with @Brennan1994 that we should defer a map window. At least we could defer it for a few mvps.

**Comment by rnugent3** [comment]
We have included in our performance work statement a task to fix bugs related to the map window and a task to view the hydraulics in the map window. We settled on the assumption that the contractor would not fix things within the mapping library, just things in the HEC-FDA codebase. I understand that we can de-obligate funds for certain tasks if needed, but if we have funds obligated for that purpose and contractor can do the work within the agreed-upon level of effort, is it worth going through the de-obligation process? What do we anticipate having to do on the map window in house?

**Comment by rnugent3** [comment]
I think we'll be in a much better place to organize thoughts about the general path forward as we get closer to the first MVP.

**Comment by HenryGeorgist** [comment]
In my opinion, we should ask first - what is the user story that drives the need for the map window, and how does it rank in the priorities of the regulatory requirements of the software. How does it rank in the pleasant use of the tool. In essesnce, what is the problem/opportunities/objectives/constraints - what are the alternatives (thanks @Brennan1994 for starting that) and what is our criteria for selection of a proposed alternative?

**Comment by rnugent3** [comment]
I understand that the user story is important for driving features and that we look to the planning process in development. I guess I am asking if this ship has sailed while we're discussing whether to set sail?

**Comment by HenryGeorgist** [comment]
Fair enough. I am pushing for what we can accomplish to meet minimum requirements by the end of FY 22, my fear is if we include everything that set sail, we may run out of wind.

**Comment by rnugent3** [comment]
Understood. The period of performance for the two tasks put the work toward the end of the contract, so I think we'll have the time to be able assess and adjust before we run out of wind.

**Comment by HenryGeorgist** [comment]
https://gdal.org/api/csharp/index.html

Apparently there is a C# bindings project on the GDAL webpage that looks to be supported.

**Reply by HenryGeorgist** [reply]
https://github.com/MaxRev-Dev/gdal.netcore
https://github.com/MaxRev-Dev/gdal.netcore/tree/master/win (for windows)

this one is .net core comparable which is pretty great news. I found this via the GDAL page referenced above, the github repo has at least 3 developers, a FAQ, unit tests, published on nuget, actively developed (last build 27 days ago), and is on the mit license.

**Comment by rnugent3** [comment, marked as answer]
We dropped the map window for 2.0 and @Brennan1994 is re-writing all geospatial processing to use the RAS API. This discussion has been resolved.
